Optimize complexity of client endpoint WeightedRoundRobinStrategy

linxGnu opened the pull request:
Motivation:

Client using EndpointGroup with current WeightedRoundRobinStrategy is slow. 
* Current algorithm complexity for selectEndpoint operation is O(N * MaxWeight) where N is number of endpoints. Arbitrary weight value (>1000 for example) cause it slow. 

Modifications:
- Sort endpoints by weight, then by host and port.
- Find the correct endpoint faster by binary search on accumulated sum of weight.
- The complexity is O(log(N)) and not depend on weight value.

Algorithm details:
- In comment code block

Result:
* Faster selectEndpoint operation.

Benchmarks:

Before
```
Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.randomMainly1Max30":
  392327.218 ±(99.9%) 9226.013 ops/s [Average]

Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.randomMax10":
  289799.381 ±(99.9%) 3761.818 ops/s [Average]

Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.randomMax100":
  75949.628 ±(99.9%) 1970.120 ops/s [Average]

Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.sameWeight":
  54412957.573 ±(99.9%) 1253690.454 ops/s [Average]

Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.unique":
  21152.272 ±(99.9%) 11486.626 ops/s [Average]
```

After
```
Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.randomMainly1Max30":
  28586820.346 ±(99.9%) 145391.823 ops/s [Average]

Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.randomMax10":
  25691640.444 ±(99.9%) 574921.960 ops/s [Average]

Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.randomMax100":
  24246013.498 ±(99.9%) 994566.750 ops/s [Average]

Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.sameWeight":
  54149030.441 ±(99.9%) 1580181.334 ops/s [Average]

Result "com.linecorp.armeria.core.client.endpoint.WeightedRoundRobinStrategyBenchmark.unique":
  21813513.256 ±(99.9%) 1958592.548 ops/s [Average]
```

imasahiro commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -71,49 +74,109 @@ public Endpoint select(ClientRequestContext ctx) {
+                int numEndpoints = endps.size();
+
+                // accumulation
+                LinkedList<EndpointsGroupByWeight> accumulatedGroups = new LinkedList<>();
```
Can we replace first `LinkedList` to `List`? And can you use `new ArrayList` instead?

linxGnu replied:
Thank you very much. I would change to use List interface.

However I think I would keep using LinkedList to prevent grow() (the nature of ArrayList does grow operation if reach capacity, which take Array.copy() operation)

trustin replied:
Perhaps you could specify a better initial capacity, so it doesn't expand too often (or even never)?

imasahiro commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -71,49 +74,109 @@ public Endpoint select(ClientRequestContext ctx) {
+                }
+
+                this.endpoints = endps;
+                this.endpointsGroupByWeight = accumulatedGroups.toArray(
```
You don't need to convert to raw array if you use `ArrayList`, I think.

linxGnu replied:
I think ArrayList is much more better in term of abstraction and grow. 

However in this PR, I just want to make this micro optimization as good as possible, not calling additional function to archive element.

I think I would count number of distinct weight and do preallocation.

anuraaga replied:
I think using an `ArrayList` will generally be faster than traversing a `LinkedList` that happens here. It seems reasonable to just allocate `new ArrayList<(endps.size())` and call `ImmutableList.copyOf` here.

Also, since this should only be called when endpoints change, which isn't often, I don't think we need to look at overoptimizations in this constructor.

imasahiro commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -71,49 +74,109 @@ public Endpoint select(ClientRequestContext ctx) {
-            private final int totalWeight;
+            private final long totalWeight; // prevent overflow by using long
+
+            private static final class EndpointsGroupByWeight {
```
Can you move this class out of `EndpointsAndWeights` for less indenting?

imasahiro commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -71,49 +74,109 @@ public Endpoint select(ClientRequestContext ctx) {
+            private final long totalWeight; // prevent overflow by using long
+
+            private static final class EndpointsGroupByWeight {
+                int startIndex;
```
Think you can add `final` to fields.

anuraaga commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -71,49 +74,109 @@ public Endpoint select(ClientRequestContext ctx) {
+                // accumulation
+                LinkedList<EndpointsGroupByWeight> accumulatedGroups = new LinkedList<>();
+                EndpointsGroupByWeight currentGroup = null;
+                int rest = numEndpoints;
```
`long rest`?

linxGnu replied:
Number of Endpoints is int (in signature of size() of List too). So I think int is ok.

anuraaga replied:
Yeah but you only use it by casting to `long` - if you use `long` here, I think you can remove all the casts.

linxGnu replied:
Yeah, and I need to change type of `startIndex` to `long` too to remove other casts.

linxGnu commented:
Thank you very much @imasahiro @anuraaga @trustin 
I have updated source code and added JMH as your suggestion. Benchmark is also included in comment.
Please take a review.

huydx commented on benchmarks/src/jmh/java/com/linecorp/armeria/core/client/endpoint/WeightedRoundRobinStrategyBenchmark.java:
```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2016 LINE Corporation
```
2018

linxGnu commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -68,26 +71,112 @@ public Endpoint select(ClientRequestContext ctx) {
+                int numberDistinctWeight = 0;
+
+                // prepare endpoints by copying
+                this.endpoints = ImmutableList.copyOf(endpoints)
```
I am sorry, I didn't notice it :(

And do you mean (I couldn't find Streams.of from Guava and JDK)
```
this.endpoints = Streams
                        .stream(endpoints)
                        .filter(e -> e.weight() > 0)
                        .sorted(Comparator
                                .comparing(Endpoint::weight)
                                .thenComparing(Endpoint::host)
                                .thenComparingInt(Endpoint::port))
                        .collect(ImmutableList.toImmutableList());
```

trustin replied:
Looks good to me. nits:

- No line break between `Streams` and `.stream(...)`
- No line break between `Comparator` and `.comparing()`
. Static-import `ImmutableList.toImmutableList`.

anuraaga replied:
I did indeed mean `Streams.stream`, sorry for the confusion!

linxGnu commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -68,26 +71,112 @@ public Endpoint select(ClientRequestContext ctx) {
+                for (Endpoint endpoint : this.endpoints) {
+                    if (currentGroup == null || currentGroup.weight != endpoint.weight()) {
+                        totalWeight += currentGroup == null ?
+                                (long) endpoint.weight() * rest
```
Thank you, I would remove it.

linxGnu commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -68,26 +71,112 @@ public Endpoint select(ClientRequestContext ctx) {
+                        currentGroup = new EndpointsGroupByWeight(
+                                numEndpoints - rest, endpoint.weight(), totalWeight
+                        );
+                        accumulatedGroups[++index] = currentGroup;
```
In this case, array give better performance. However, to fully guarantee immutability, I would change it to `ImmutableList` using `ImmutableList.Builder` as your suggestion.

hyangtack commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -68,26 +72,112 @@ public Endpoint select(ClientRequestContext ctx) {
+                                .thenComparing(Endpoint::host)
+                                .thenComparingInt(Endpoint::port))
+                        .collect(toImmutableList());
+                long numEndpoints = this.endpoints.size();
```
nit: would you add `final` keyword to the variable which holds the value not to be changed like here?

linxGnu replied:
I would change and commit soon. Thank you very much.

trustin commented:
Thanks you, @linxGnu and reviewers. :smile:
